polar-core/src/vm.rs

```diff
@@ -650,9 +683,11 @@ impl PolarVirtualMachine {
 
     /// Push a binding onto the binding stack.
     pub fn bind(&mut self, var: &Symbol, val: Term) -> Result<()> {
-        if self.log {
-            self.print(&format!("⇒ bind: {} ← {}", var.to_polar(), val.to_polar()));
-        }
+        self.log_with(
+            LogLevel::Trace,
+            || format!("⇒ bind: {} ← {}", var.to_polar(), val.to_polar()),
```

Comment:
nit: there are a bunch of `format!("{}", x.to_polar())` expressions across these `log_with()` calls where the `.to_polar()` call is redundant since the `.to_string()` impl shells out to `.to_polar()`

Reply:
cleaned up `.to_polar` calls and added some `fmt::Display` impls to fill in the gaps.